src/compute-client/src/plan/mod.rs

```diff
@@ -1373,8 +1383,17 @@ impl<T: timely::progress::Timestamp> Plan<T> {
 Dataflow info: {}
 This is not expected to cause incorrect results, but could indicate a performance issue in Materialize.", missing, debug_info);
                         } else {
-                            // It's fine and expected that linear joins don't have all their arrangements available up front,
-                            // so no need to print an error here.
+                            soft_panic_or_log!("Arrangements depended on by a non-delta join are absent: {:?}
```

Comment:
You can change the comment above as well:

```
// join_implementation.rs produced a sub-optimal plan here;
// we shouldn't plan delta joins at all if not all of the required arrangements
// are available. Soft panic in CI and log an error in production to increase the chances 
// that we will catch all situations that violate this constraint.
```

Reply:
Changed